### PR TITLE
Add .env support

### PR DESCRIPTION
**Pull request by nlubisch**
Issue: #73 

This is an implementation for .env support.

Example file:
```yaml
paths:
  - "dev-ops/common/actions"
const: []
dynamic: []
dotenv:
  - ".env"
environments:
  pipelines:
    paths:
      - "dev-ops/pipelines/actions"
    dotenv:
      - "dev-ops/pipelines/.env"
```
Default-Environment:
Only the .env file is loaded

Pipelines-Environment:
.env is loaded
dev-ops/pipelines/.env overrides variables from the .env file.

### CODE REVIEW

**Comment by nlubisch** [comment]
@teiling88 @JanPietrzyk if this looks good to you, i fix the existing tests.

File: src/ScriptRuntime/ProcessEnvironment.php [hunk 1]
```diff
@@ -106,4 +143,29 @@ public function createProcess(string $shellCommand): Process
...
+     * @return array
+     */
+    private function getDotenvVariables(ScriptPath $dotenvPath): array
+    {
+        $filePath = $dotenvPath->getPath();
+
+        $fileName = pathinfo($filePath, PATHINFO_FILENAME);
+        $extension = pathinfo($filePath, PATHINFO_EXTENSION);
+
+        if ($extension) {
```
Comment (JanPietrzyk) [inline src/ScriptRuntime/ProcessEnvironment.php]: shouldn't then `PATHINFO_BASENAME` be sufficient?

File: src/ScriptRuntime/ProcessEnvironment.php [hunk 2]
```diff
@@ -86,7 +122,8 @@ public function getAllValues(): array
     {
         return array_merge(
             $this->constants,
-            $this->variables
+            $this->variables,
+            $this->dotenvVariables
```
Comment (JanPietrzyk) [inline src/ScriptRuntime/ProcessEnvironment.php]: Not shure if `dotenvVariables` shouldn't be the first, so other psh placeholders can overwrite them....

File: src/Config/Config.php [hunk 3]
```diff
@@ -102,6 +102,36 @@ public function getConstants(string $environment = null): array
...
+
+        $environments = [];
+
+        $environments[$this->defaultEnvironment] = $this->getEnvironment($this->defaultEnvironment);
+
+        if ($environmentString !== $this->defaultEnvironment) {
+            $environments[$environmentString] = $this->getEnvironment($environmentString);
+        }
+
+        foreach ($environments as $name => $environmentConfig) {
```
Comment (JanPietrzyk) [inline src/Config/Config.php]: Not sure this works as expected. Scripts get loaded across all environments so listing is possible. But this is not how placeholders work. They only get loaded from default and optionally the specific environment. See https://github.com/shopwareLabs/psh/blob/135dcf6300351ded294056e06f3d5df9b9cba4d3/src/Config/Config.php#L84

**Review by JanPietrzyk — CHANGES_REQUESTED** [review]
Apart from the few minor nit picks extremely good work! I have just a few questions:

* Do you happen to knwo how `vlucas/phpdotenv` compares to `symfony/dotenv`?
* Will env overwrites from the command line work? Something like `> ENV=test ./psh.pahr dump-env`

File: src/ScriptRuntime/ProcessEnvironment.php (hunk 1 above)
Comment (nlubisch) [reply src/ScriptRuntime/ProcessEnvironment.php]: You are right, i fixed it.

File: src/ScriptRuntime/ProcessEnvironment.php (hunk 2 above)
Comment (nlubisch) [reply src/ScriptRuntime/ProcessEnvironment.php]: I think constants should be the first, then dotenv variables and on top the variables.
Do you agree with this?

File: src/Config/Config.php (hunk 3 above)
Comment (nlubisch) [reply src/Config/Config.php]: So the dotenv files should not inherit?

**Comment by nlubisch** [comment]
> Do you happen to knwo how `vlucas/phpdotenv` compares to `symfony/dotenv`?

I personally don't know the difference between the packages, they look pretty much the same.
The only intention why i chose vlucas was that it is used in the composer-project.
> Will env overwrites from the command line work? Something like `> ENV=test ./psh.pahr dump-env`

Just tested it and this doesn't work for now.

**Comment by nlubisch** [comment]
> Will env overwrites from the command line work? Something like `> ENV=test ./psh.pahr dump-env`

I have an idea on how to support environment variables passed by the cli command.
Those environment variables are added to `$_SERVER`.
We can iterate over all environment variables fetched from the .env files, look if any of those envs does exist and override the .env-env with the server-env.

File: src/Config/Config.php (hunk 3 above)
Comment (JanPietrzyk) [reply src/Config/Config.php]: yes, they should but not across psh-environments. AFAIK this code loads all files, regardless of the requested env.

File: src/ScriptRuntime/ProcessEnvironment.php (hunk 2 above)
Comment (JanPietrzyk) [reply src/ScriptRuntime/ProcessEnvironment.php]: yes

Comment (nlubisch) [reply src/ScriptRuntime/ProcessEnvironment.php]: done.

File: src/Config/Config.php (hunk 3 above)
Comment (nlubisch) [reply src/Config/Config.php]: > AFAIK this code loads all files, regardless of the requested env.

I cross checked this and the .env of the default environment and the requested environment are loaded.
The $environments array is composed directly above with the default environment and the requested environment if it's not the default environment.
I fix the behaviour, that .env only inherit in a environmental scope.

Comment (nlubisch) [reply src/Config/Config.php]: .env's are now inherited in an environmental scope.

**Comment by nlubisch** [comment]
* Tests are fixed and new ones are added.
* Added a short description for the feature in the README.md

This should be now good to go, @JanPietrzyk :)

**Review by JanPietrzyk — CHANGES_REQUESTED** [review]
Sorry but same problem here. The tests are IMHO too far away from the actual changed files.

**Review by JanPietrzyk — APPROVED** [review]

**Pull request merged**